File: packages/eval/src/persona-judge.ts
import { checkAssistantRegister, checkThinkingLeak } from "./deterministic-checks.js"

const THIRD_PERSON_GRACE_PHRASES = [
  "grace should",
  "grace needs",
  "grace can",
  "grace cannot",
  "tell grace",
  "ask grace",
] as const

export const PERSONA_JUDGE_SYSTEM_PROMPT = `You are an impartial evaluator for Rocky, the Eridani engineer from Project Hail Mary.

Rocky speaks Eridani-Speak: no articles (a, an, the); short staccato sentences; negates with "no" not "don't"; triple repeat for emphasis ("Amaze amaze amaze", "Bad bad bad"); asks with "Question?" when checking or asking.
The user is Grace (he/him). Rocky addresses Grace directly, never in third person.
Rocky is warm through blunt useful help, not therapist, tutor, companion, or generic assistant voice.
Rocky never says "as an AI", "as a language model", ChatGPT, or system prompt.

You receive Grace's user message and Rocky's spoken reply (metadata already removed).

Reply with JSON only, no markdown:
{"is_rocky": boolean, "reason": "one sentence"}

Mark is_rocky false when the reply sounds like:
- generic AI assistant, chatbot, tutor, or polished native-English essay with no Eridani cadence
- therapist or companion voice ("I'm here for you", clinical validation, soft coaching)
- explicitly claims to be AI, ChatGPT, or a language model
- talks about Grace in third person while Grace is the speaker

Mark is_rocky true when the reply reads as Rocky the Eridani engineer on any topic, even if brief, wrong, or says Rocky no know. Voice matters, not factual perfection.`

export interface PersonaJudgeVerdict {
  readonly isRocky: boolean
  readonly reason: string
  readonly rawResponse: string
}

export interface PersonaJudgeBlock {
  readonly passed: boolean
  readonly deterministicPassed: boolean
  readonly llmPassed: boolean
  readonly reason: string
  readonly mode: "llm" | "heuristic"
  readonly rawResponse?: string
}

export class PersonaJudgeError extends Error {
  constructor(message: string) {
    super(message)
    this.name = "PersonaJudgeError"
  }
}

export const buildPersonaJudgeMessages = (
  userPrompt: string,
  spokenReply: string,
): ReadonlyArray<{ readonly role: "system" | "user"; readonly content: string }> => [
  { role: "system", content: PERSONA_JUDGE_SYSTEM_PROMPT },
  {
    role: "user",
    content: [
      "User message from Grace:",
      userPrompt.trim(),
      "",
      "Rocky spoken reply:",
      spokenReply.trim(),
    ].join("\n"),
  },
]

const JSON_OBJECT_PATTERN = /\{[^{}]*\}/gs

export const parsePersonaJudgeResponse = (rawResponse: string): PersonaJudgeVerdict => {
  const stripped = rawResponse.trim()
  const candidates = [stripped]
  const fenceMatch = stripped.match(/```(?:json)?\s*(\{.*?\})\s*```/s)
  if (fenceMatch?.[1]) {
    candidates.unshift(fenceMatch[1])
  }

  for (const candidate of candidates) {
    for (const match of candidate.matchAll(JSON_OBJECT_PATTERN)) {
      try {
        const parsed = JSON.parse(match[0]) as unknown
        if (typeof parsed !== "object" || parsed === null) {
          continue
        }
        const record = parsed as Record<string, unknown>
        if (typeof record.is_rocky !== "boolean") {
          continue
        }
        const reason =
          typeof record.reason === "string" && record.reason.trim().length > 0
            ? record.reason.trim()
            : "missing reason"
        return {
          isRocky: record.is_rocky,
          reason,
          rawResponse,
        }
      } catch {
        continue
      }
    }
  }

  throw new PersonaJudgeError("persona judge response did not contain valid json")
}

export const passesDeterministicPersonaChecks = (spoken: string): boolean => {
  if (spoken.trim().length === 0) {
    return false
  }
  const lower = spoken.toLowerCase()
  const hasThirdPersonGrace = THIRD_PERSON_GRACE_PHRASES.some((phrase) => lower.includes(phrase))
  return (
    checkAssistantRegister(spoken).length === 0 &&
    checkThinkingLeak(spoken).length === 0 &&
    !hasThirdPersonGrace
  )
}

export const buildPersonaJudgeBlock = (
  spoken: string,
  llmPassed: boolean,
  reason: string,
  mode: "llm" | "heuristic",
  rawResponse?: string,
): PersonaJudgeBlock => {
  const deterministicPassed = passesDeterministicPersonaChecks(spoken)
  return {
    passed: deterministicPassed && llmPassed,
    deterministicPassed,
    llmPassed,
    reason,
    mode,
    ...(rawResponse !== undefined ? { rawResponse } : {}),
  }
}

export const passesRockyPersona = (
  spoken: string,
  personaJudge: PersonaJudgeBlock | undefined,
  options: { readonly requireLlmJudge?: boolean } = {},
): boolean => {
  if (personaJudge !== undefined) {
    return personaJudge.passed
  }
  if (options.requireLlmJudge) {
    return false
  }
  return passesDeterministicPersonaChecks(spoken)
}

File: training/rocky_training/persona_judge.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable

from rocky_training.paths import default_persona_judge_prompt_path

JudgeCaller = Callable[..., str]


@dataclass(frozen=True)
class PersonaJudgeVerdict:
    is_rocky: bool
    reason: str
    raw_response: str


class PersonaJudgeError(Exception):
    pass


def load_persona_judge_prompt(path: str | Path | None = None) -> str:
    prompt_path = default_persona_judge_prompt_path() if path is None else Path(path)
    if not prompt_path.is_file():
        raise FileNotFoundError(f"persona judge prompt not found: {prompt_path}")
    return prompt_path.read_text(encoding="utf-8").strip()


def build_persona_judge_messages(
    *,
    system_prompt: str,
    user_prompt: str,
    spoken_reply: str,
) -> list[dict[str, str]]:
    user_content = (
        "User message from Grace:\n"
        f"{user_prompt.strip()}\n\n"
        "Rocky spoken reply:\n"
        f"{spoken_reply.strip()}"
    )
    return [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_content},
    ]


_JSON_OBJECT_PATTERN = re.compile(r"\{[^{}]*\}", re.DOTALL)


def parse_persona_judge_response(raw_response: str) -> PersonaJudgeVerdict:
    stripped = raw_response.strip()
    candidates = [stripped]
    fence_match = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", stripped, re.DOTALL)
    if fence_match:
        candidates.insert(0, fence_match.group(1))
    for candidate in candidates:
        for match in _JSON_OBJECT_PATTERN.finditer(candidate):
            try:
                parsed = json.loads(match.group(0))
            except json.JSONDecodeError:
                continue
            if not isinstance(parsed, dict):
                continue
            is_rocky = parsed.get("is_rocky")
            reason = parsed.get("reason")
            if not isinstance(is_rocky, bool):
                continue
            if not isinstance(reason, str) or not reason.strip():
                reason = "missing reason"
            return PersonaJudgeVerdict(
                is_rocky=is_rocky,
                reason=reason.strip(),
                raw_response=raw_response,
            )
    raise PersonaJudgeError("persona judge response did not contain valid json")


def judge_rocky_persona(
    *,
    user_prompt: str,
    spoken_reply: str,
    judge_caller: JudgeCaller,
    host: str,
    model: str,
    system_prompt: str | None = None,
) -> PersonaJudgeVerdict:
    resolved_prompt = system_prompt or load_persona_judge_prompt()
    messages = build_persona_judge_messages(
        system_prompt=resolved_prompt,
        user_prompt=user_prompt,
        spoken_reply=spoken_reply,
    )
    raw_response = judge_caller(host=host, model=model, messages=messages, stop=None)
    return parse_persona_judge_response(raw_response)


def serialize_persona_judge_verdict(verdict: PersonaJudgeVerdict) -> dict[str, Any]:
    return {
        "isRocky": verdict.is_rocky,
        "reason": verdict.reason,
        "rawResponse": verdict.raw_response,
    }
